fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -660,6 +680,197 @@ public void exec() throws Exception {
+                            backendExecStates.put(tUniquePrams.backend_num, execState);
+                            if (needCheckBackendState) {
+                                needCheckBackendExecStates.add(execState);
+                                if (LOG.isDebugEnabled()) {
+                                    LOG.debug("add need check backend {} for fragment, {} job: {}",
+                                            execState.backend.getId(),
+                                            fragment.getFragmentId().asInt(), jobId);
+                                }
+                            }
+                            lastExecState = execState;
+                        }
+
+                        futures.add(Pair.create(lastExecState, lastExecState.execRemoteMultiFragmentsAsync(tRequest)));
+                    }
+
+                    profileFragmentId += 1;
+                }
+
+                for (Pair<BackendExecState, Future<PExecMultiPlanFragmentsResult>> pair : futures) {
+                    TStatusCode code;
+                    String errMsg = null;
+                    try {
+                        PExecMultiPlanFragmentsResult result =
+                                pair.second.get(queryDeliveryTimeoutMs, TimeUnit.MILLISECONDS);
+                        code = TStatusCode.findByValue(result.status.statusCode);
+                        if (result.status.errorMsgs != null && !result.status.errorMsgs.isEmpty()) {
+                            errMsg = result.status.errorMsgs.get(0);
+                        }
+                    } catch (ExecutionException e) {
+                        LOG.warn("catch a execute exception", e);
+                        code = TStatusCode.THRIFT_RPC_ERROR;
+                    } catch (InterruptedException e) {
+                        LOG.warn("catch a interrupt exception", e);
+                        code = TStatusCode.INTERNAL_ERROR;
+                    } catch (TimeoutException e) {
+                        LOG.warn("catch a timeout exception", e);
+                        code = TStatusCode.TIMEOUT;
+                    }
+
+                    if (code != TStatusCode.OK) {
+                        if (errMsg == null) {
+                            errMsg = "exec rpc error. backend id: " + pair.first.backend.getId();
+                        }
+                        queryStatus.setStatus(errMsg);
+                        LOG.warn("exec plan fragment failed, errmsg={}, code: {}, fragmentId={}, backend={}:{}",
+                                errMsg, code, pair.first.fragmentId,
+                                pair.first.address.hostname, pair.first.address.port);
+                        cancelInternal(PPlanFragmentCancelReason.INTERNAL_ERROR);
+                        switch (Objects.requireNonNull(code)) {
+                            case TIMEOUT:
+                                throw new UserException("query timeout. backend id: " + pair.first.backend.getId());
```

Comment:
Maybe needn't, in this way the query could be canceled as fast as possible, and this is the same behaviour as the current delivery way.
If some RPC response a failed status,
- if the connection has been closed by FE, the subsequent responses will fail.
- otherwise, the response will be ignored.